Libraries/libutil/Sources/DefaultFilesystem.cpp

```diff
@@ -51,6 +52,168 @@ isExecutable(std::string const &path) const
+PermissionsMode(Permissions permissions)
+{
+    mode_t mode = 0;
+    mode |= (permissions.flag(Permissions::Flag::Sticky) ? S_ISVTX : 0);
+    mode |= (permissions.flag(Permissions::Flag::SetUserID) ? S_ISUID : 0);
+    mode |= (permissions.flag(Permissions::Flag::SetGroupID) ? S_ISUID : 0);
+    mode |= (permissions.user(Permissions::Permission::Read) ? S_IRUSR : 0);
+    mode |= (permissions.user(Permissions::Permission::Write) ? S_IWUSR : 0);
+    mode |= (permissions.user(Permissions::Permission::Execute) ? S_IXUSR : 0);
+    mode |= (permissions.group(Permissions::Permission::Read) ? S_IRUSR : 0);
+    mode |= (permissions.group(Permissions::Permission::Write) ? S_IWUSR : 0);
+    mode |= (permissions.group(Permissions::Permission::Execute) ? S_IXUSR : 0);
+    mode |= (permissions.other(Permissions::Permission::Read) ? S_IRUSR : 0);
+    mode |= (permissions.other(Permissions::Permission::Write) ? S_IWUSR : 0);
+    mode |= (permissions.other(Permissions::Permission::Execute) ? S_IXUSR : 0);
+    return mode;
+}
+
+bool DefaultFilesystem::
+writeFilePermissions(std::string const &path, Permissions::Operation operation, Permissions permissions)
+{
+    ext::optional<Permissions> updated = this->readFilePermissions(path);
+    if (!updated) {
+        return false;
+    }
+
+    updated->combine(operation, permissions);
+    mode_t mode = PermissionsMode(*updated);
+
+    if (::chmod(path.c_str(), mode) != 0) {
+        return false;
+    }
+
+    return true;
+}
+
+bool DefaultFilesystem::
+writeSymbolicLinkPermissions(std::string const &path, Permissions::Operation operation, Permissions permissions)
+{
+#if defined(__APPLE__) || defined(__FreeBSD__)
+    ext::optional<Permissions> updated = this->readSymbolicLinkPermissions(path);
+    if (!updated) {
+        return false;
+    }
+
+    updated->combine(operation, permissions);
+    mode_t mode = PermissionsMode(*updated);
+
+    if (::lchmod(path.c_str(), mode) != 0) {
+        return false;
+    }
+
+    return true;
+#else
+    /* Symbolic links have no permissions on most filesystems. */
+    return true;
+#endif
+}
+
+bool DefaultFilesystem::
+writeDirectoryPermissions(std::string const &path, Permissions::Operation operation, Permissions permissions, bool recursive)
+{
+    ext::optional<Permissions> updated = this->readDirectoryPermissions(path);
+    if (!updated) {
+        return false;
+    }
+
+    updated->combine(operation, permissions);
+    mode_t mode = PermissionsMode(*updated);
+
+    if (::chmod(path.c_str(), mode) != 0) {
+        return false;
+    }
+
+    if (recursive) {
+        bool success = true;
+
+        success &= this->readDirectory(path, recursive, [this, &path, &operation, &permissions, &success](std::string const &name) {
+            std::string full = path + "/" + name;
+            if (this->isDirectory(full)) {
+                if (!this->writeDirectoryPermissions(full, operation, permissions, false)) {
+                    success = false;
```

Comment:
you don't need to assign success as false, do you?  If you return 'false', and you're &ing the value to success, it'll get updated anyways.